src/datasources/notifications/notifications.datasource.spec.ts

```diff
@@ -0,0 +1,633 @@
+                  chain_id: safe.chainId,
+                  safe_address: safe.address,
+                  created_at: expect.any(Date),
+                  updated_at: expect.any(Date),
+                };
+              }),
+            ),
+        );
+        expect(subscribedNotifications).toStrictEqual(
+          expect.arrayContaining(
+            upsertSubscriptionsDto.safes
+              .flatMap((safe, i) => {
+                return safe.notificationTypes.map((type) => {
+                  return {
+                    id: expect.any(Number),
+                    notification_subscription_id: i + 1,
+                    notification_type_id: types.find((t) => t.name === type)
+                      ?.id,
+                  };
+                });
+              })
+              .concat(
+                upsertSubscriptionsDto.safes.flatMap((safe, i) => {
+                  return safe.notificationTypes.map((type) => {
+                    return {
+                      id: expect.any(Number),
+                      notification_subscription_id:
+                        upsertSubscriptionsDto.safes.length + i + 1,
+                      notification_type_id: types.find((t) => t.name === type)
+                        ?.id,
+                    };
+                  });
+                }),
+              ),
+          ),
+        );
+      });
+    });
+  });
+
+  describe('getSafeSubscription', () => {
+    it('should return a subscription for a Safe', async () => {
+      const signerAddress = getAddress(faker.finance.ethereumAddress());
+      const upsertSubscriptionsDto = upsertSubscriptionsDtoBuilder().build();
+      await target.upsertSubscriptions({
+        signerAddress,
+        upsertSubscriptionsDto,
+      });
+
+      const safe = upsertSubscriptionsDto.safes[0];
+      await expect(
+        target.getSafeSubscription({
+          signerAddress,
+          deviceUuid: upsertSubscriptionsDto.deviceUuid!,
+          chainId: safe.chainId,
+          safeAddress: safe.address,
+        }),
+      ).resolves.toStrictEqual(expect.arrayContaining(safe.notificationTypes));
+    });
+  });
+
+  describe('getSubscribersWithTokensBySafe', () => {
+    it('should return a list of subscribers with tokens for a Safe', async () => {
+      const signerAddress = getAddress(faker.finance.ethereumAddress());
+      const upsertSubscriptionsDto = upsertSubscriptionsDtoBuilder().build();
+      const secondSignerAddress = getAddress(faker.finance.ethereumAddress());
+      const secondUpsertSubscriptionsDto = upsertSubscriptionsDtoBuilder()
+        .with('safes', upsertSubscriptionsDto.safes)
+        .build();
+      await target.upsertSubscriptions({
+        signerAddress,
+        upsertSubscriptionsDto,
+      });
+      await target.upsertSubscriptions({
+        signerAddress: secondSignerAddress,
+        upsertSubscriptionsDto: secondUpsertSubscriptionsDto,
+      });
```

Comment:
Nit: could we add a third subscription with different `safes`, so we verify it's not included in the result?

Reply:
Added in d49aca2e6a097d8ac438fa5876d39f1b2006bf3f.